internal/indexcoord/node_manager.go

```diff
@@ -169,17 +190,20 @@ type indexNodeGetMetricsResponse struct {
 
 // getMetrics get metrics information of all IndexNode.
 func (nm *NodeManager) getMetrics(ctx context.Context, req *milvuspb.GetMetricsRequest) []indexNodeGetMetricsResponse {
+	var clients []types.IndexNode
 	nm.lock.RLock()
-	defer nm.lock.RUnlock()
+	for _, node := range nm.nodeClients {
+		clients = append(clients, node)
+	}
+	nm.lock.RUnlock()
 
 	ret := make([]indexNodeGetMetricsResponse, 0, len(nm.nodeClients))
-	for _, node := range nm.nodeClients {
+	for _, node := range clients {
 		resp, err := node.GetMetrics(ctx, req)
```

Comment:
It feels weird that `ListNodes` uses `nm.ctx` while `getMetrics` uses parameter context. Also `ListNodes` request `IndexNode` in parallel while `getMetrics` in sequence.

Reply:
@congqixia 

getMetrics defined in IndexCoord use parameter context is to let the external request context to control when to cancel the request.

getMetrics called in ListNode is an internal request  to IndexNode, and is controlled by the ctx  derived from nm.ctx (which is IndexCoord 's loopCtx) through context.WithTimeout. I choose the nm.ctx instead of context.Background(). Currently, we don't  create and store contexts related to separate IndexNode  in IndexCoord.

Indeed, getMetrics  can be parallel, I need to add another local Sync.Mutex to control the append of slice.  External call of getMetrics is infrequent. I think this can be improved in the future.

Reply:
If you insist on parallelism in GetMetrics , I will add the logic. ^_^

Reply:
We will parallelize all the GetMetrics in the future.